### PR TITLE
Refine Parquet schema mismatch message

### PR DESCRIPTION
**Pull request by zhenxiao**
@nezihyigitbasi

### CODE REVIEW

File: presto-parquet/src/main/java/com/facebook/presto/parquet/reader/ParquetReader.java [hunk 1]
```diff
@@ -206,7 +210,16 @@ private ColumnChunk readPrimitive(PrimitiveField field)
             ParquetColumnChunk columnChunk = new ParquetColumnChunk(descriptor, buffer, 0);
             columnReader.setPageReader(columnChunk.readAllPages());
         }
-        return columnReader.readPrimitive(field);
+        try {
+            return columnReader.readPrimitive(field);
+        }
+        catch (UnsupportedOperationException e) {
+            throw new PrestoException(INVALID_SCHEMA_PROPERTY, format("There is a mismatch between file schema and partition schema. The column %s in file %s is declared as type %s but parquet file declared column type as %s.",
```
Comment (nezihyigitbasi) [inline presto-parquet/src/main/java/com/facebook/presto/parquet/reader/ParquetReader.java]: How about:

```
The column %s is declared as type %s, but the Parquet file %s declares the column as type %s
```

File: presto-parquet/src/main/java/com/facebook/presto/parquet/reader/ParquetReader.java [hunk 2]
```diff
@@ -206,7 +210,16 @@ private ColumnChunk readPrimitive(PrimitiveField field)
             ParquetColumnChunk columnChunk = new ParquetColumnChunk(descriptor, buffer, 0);
             columnReader.setPageReader(columnChunk.readAllPages());
         }
-        return columnReader.readPrimitive(field);
+        try {
+            return columnReader.readPrimitive(field);
+        }
+        catch (UnsupportedOperationException e) {
```
Comment (nezihyigitbasi) [inline presto-parquet/src/main/java/com/facebook/presto/parquet/reader/ParquetReader.java]: `UnsupportedOperationException` is a broad exception to catch here. Maybe we should throw a specific Parquet exception and catch it here. What do you think?

File: presto-parquet/src/main/java/com/facebook/presto/parquet/reader/ParquetReader.java (hunk 1 above)
Comment (zhenxiao) [reply presto-parquet/src/main/java/com/facebook/presto/parquet/reader/ParquetReader.java]: sure

File: presto-parquet/src/main/java/com/facebook/presto/parquet/reader/ParquetReader.java (hunk 2 above)
Comment (zhenxiao) [reply presto-parquet/src/main/java/com/facebook/presto/parquet/reader/ParquetReader.java]: sure. UnsupportedOperationException is actually from type.writeLong etc, when lower level ColumnReader trying to read parquet values. How about throw ParquetCorruptionException there, and catch it here?

**Comment by zhenxiao** [comment]
@nezihyigitbasi I get comments addressed
could you please review?

**Comment by zhenxiao** [comment]
kindly ping @nezihyigitbasi

**Comment by nezihyigitbasi** [comment]
@arhimondr @rschlussel can you please take a look?

**Comment by arhimondr** [comment]
From what i understand, this PR tries to address the confusing error message in case there's a schema mismatch between a file and the partition schema.

@zhenxiao Could you please elaborate a little bit more why it was decided to implement it this way? (by catching the exception in the very last moment). 

For example the schema validation can be done early, when getting the Parquet Type: 
https://github.com/prestodb/presto/blob/master/presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java#L163
https://github.com/prestodb/presto/blob/master/presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java#L255
The HiveColumnHandle contains the Hive Schema type, so it is possible to check if types match.

@nezihyigitbasi Do you know how something similar is done in ORC?

**Comment by zhenxiao** [comment]
thank you, @arhimondr 
get comments addressed. how about this approach?

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java [hunk 3]
```diff
@@ -251,13 +269,45 @@ public static ParquetPageSource createParquetPageSource(
...
+                                            column.getHiveType(),
+                                            type.asPrimitiveType().getPrimitiveTypeName()));
+        }
+        return type;
+    }
+
+    private static boolean schemaMatch(org.apache.parquet.schema.Type parquetType, HiveColumnHandle column)
+    {
+        String prestoType = column.getTypeSignature().getBase();
+        if (prestoType.equals(MAP) || prestoType.equals(ARRAY) || prestoType.equals(ROW)) {
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java]: Could you please check the signature for the complex types recursively?

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java [hunk 4]
```diff
@@ -251,13 +269,45 @@ public static ParquetPageSource createParquetPageSource(
...
+            return true;
+        }
+        if (parquetType.isPrimitive()) {
+            PrimitiveTypeName parquetTypeName = parquetType.asPrimitiveType().getPrimitiveTypeName();
+            return ((parquetTypeName == PrimitiveTypeName.INT32 && (prestoType.equals(INTEGER) || prestoType.equals(SMALLINT) || prestoType.equals(DATE) || prestoType.equals(DECIMAL) || prestoType.equals(TINYINT))) ||
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java]: nit: switch seems to be more readable

```
switch (parquetTypeName) {
                case INT64:
                    return prestoType.equals(BIGINT) || prestoType.equals(DECIMAL);
                case INT32:
                    return prestoType.equals(INTEGER) || prestoType.equals(SMALLINT) || prestoType.equals(DATE) || prestoType.equals(DECIMAL) || prestoType.equals(TINYINT);
                case BOOLEAN:
                    return prestoType.equals(BOOLEAN);
                case BINARY:
                    return prestoType.equals(VARBINARY) || prestoType.equals(VARCHAR) || prestoType.startsWith(CHAR);
                case FLOAT:
                    return prestoType.equals(REAL);
                case DOUBLE:
                    return prestoType.equals(DOUBLE);
                case INT96:
                    return prestoType.equals(TIMESTAMP);
                case FIXED_LEN_BYTE_ARRAY:
                    return prestoType.equals(DECIMAL);
                default:
                    throw new IllegalArgumentException("Unexpected parquet type name: " + parquetTypeName);
            }
```

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java (hunk 3 above)
Comment (arhimondr) [reply presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java]: I would suggest passing here the resolved `Type` object, so the convenient `Type#getTypeParameters` can be used to recurse into the nested types.

Comment (arhimondr) [reply presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java]: e.g.: https://github.com/prestodb/presto/blob/e313922f234948bfba27f9334905a750580d824f/presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSource.java#L98

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveFileFormats.java [hunk 5]
```diff
@@ -510,6 +516,67 @@ public void testFailForLongVarcharPartitionColumn()
                 .isFailingForRecordCursor(new GenericHiveRecordCursorProvider(HDFS_ENVIRONMENT), expectedErrorCode, expectedMessage);
     }
 
+    @Test
+    public void testSchemaMismatch()
+            throws Exception
+    {
+        TestColumn floatColumn = new TestColumn("column_name", javaFloatObjectInspector, 5.1f, 5.1f);
```
Comment (arhimondr) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveFileFormats.java]: Could you please add test cases for the complex types?

**Comment by zhenxiao** [comment]
thank you, @arhimondr 
I get comments addressed
I just found, actually, the schema checking should only work for Parquet index based access, for Parquet name based access, the fields inside structs could change order, or primitive columns could change order, should not check schema in this case.
Could you please review when u are free?

**Pull request merged**